File: drills/ts/solutions/nc24_trapping_rain_water__prefix_maxima.ts
export function trap(height: number[]): number {
  const left: number[] = [];
  let best = 0;
  for (const h of height) {
    best = Math.max(best, h);
    left.push(best);
  }

  const right = new Array<number>(height.length);
  best = 0;
  for (let i = height.length - 1; i >= 0; i--) {
    best = Math.max(best, height[i]);
    right[i] = best;
  }

  return height.reduce((total, h, i) => total + Math.min(left[i], right[i]) - h, 0);
}

File: drills/go/solutions/nc24_trapping_rain_water__prefix_maxima.go
package main

func trap(height []int) int {
	n := len(height)
	if n == 0 {
		return 0
	}
	// Water above a bar is bounded by the tallest bar to each side.
	leftMax := make([]int, n)
	rightMax := make([]int, n)
	leftMax[0] = height[0]
	for i := 1; i < n; i++ {
		leftMax[i] = max(leftMax[i-1], height[i])
	}
	rightMax[n-1] = height[n-1]
	for i := n - 2; i >= 0; i-- {
		rightMax[i] = max(rightMax[i+1], height[i])
	}
	water := 0
	for i, h := range height {
		water += min(leftMax[i], rightMax[i]) - h
	}
	return water
}
